lambda-buffers-compiler/src/LambdaBuffers/Compiler/TypeClass/Solve.hs

```diff
@@ -0,0 +1,135 @@
+getSubs (ProdP xs) (ProdP xs') = getSubs xs xs'
+getSubs (RecP xs) (RecP xs') = getSubs xs xs'
+getSubs (SumP xs) (SumP xs') = getSubs xs xs'
+getSubs (AppP t1 t2) (AppP t1' t2') = getSubs t1 t1' <> getSubs t2 t2'
+getSubs (RefP n t) (RefP n' t') = getSubs n n' <> getSubs t t'
+getSubs (DecP a b c) (DecP a' b' c') = getSubs a a' <> getSubs b b' <> getSubs c c'
+getSubs _ _ = []
+
+normalize :: Ord a => [a] -> [a]
+normalize = S.toList . S.fromList
+
+-- TODO: integrate the specificity stuff to make `solve` a total function
+isSubstitutionOf :: Pat -> Pat -> Bool
+isSubstitutionOf = flip matches
+
+compareSpecificity :: Pat -> Pat -> Ordering
+compareSpecificity p1 p2
+  | p1
+      `isSubstitutionOf` p2
+      && p2
+      `isSubstitutionOf` p1 =
+      EQ
+  | p1 `isSubstitutionOf` p2 = LT
+  | otherwise = GT
+
+sortOnSpecificity :: Pat -> [Pat] -> [Pat]
+sortOnSpecificity p ps = sortBy compareSpecificity $ filter (`matches` p) ps
+
+mostSpecificInstance :: Pat -> [Pat] -> Maybe Pat
+mostSpecificInstance p ps = case sortOnSpecificity p ps of
+  [] -> Nothing
+  (x : _) -> Just x
+
+{- Given a list of instances (the initial scope), determines whether we can derive
+   an instance of the Class argument for the Pat argument. A result of [] indicates that there are
+   no remaining subgoals and that the constraint has been solved.
+
+   NOTE: At the moment this handles superclasses differently than you might expect -
+         instead of assuming that the superclasses for all in-scope classes are defined,
+         we check that those constraints can be solved before affirmatively judging that the
+         target constraint has been solved. I *think* that makes sense in this context (whereas in Haskell
+         it doesn't b/c it's *impossible* to have `instance Foo X` if the definition of Foo is
+         `class Bar y => Foo y` without an `instance Bar X`)
+
+   TODO: This needs *extensive* testing, which is somewhat complicated to implement.
+-}
+solve ::
+  [Instance] -> -- all instances in scope. WE ASSUME THESE HAVE ALREADY BEEN GENERATED, SOMEWHERE
+  Constraint ->
+  [Constraint]
+solve inScope cst@(C c pat) =
+  case flip subst pat <$> matchHeads inScope of
+    [] -> [cst]
+    [C _ p :<= []] -> case supers c of
+      [] -> []
+      xs -> normalize . concat $ traverse (solve inScope . flip C p) xs
+    [C _ _ :<= is] -> case normalize $ goConstraints is of
+      [] -> normalize . concat $ traverse (solve inScope . flip C pat) (supers c)
+      xs -> xs
+    xs -> error ("Multiple matches, coherence broken: " <> show xs) -- TODO: Use specificity stuff to eliminate this branch
+  where
+    goConstraints :: [Constraint] -> [Constraint]
+    goConstraints [] = []
+    goConstraints (cx : rest) =
+      let xs = solve inScope cx
+       in xs <> goConstraints rest
+
+    matchHeads :: [Instance] -> [Instance]
+    matchHeads xs = flip filter xs $ \(C c' t' :<= _) -> c == c' && matches t' pat
```

Comment:
Please give some named let bindings for the reader and abandon the extensive use of flip

Reply:
I'm not really sure where you want `let` bindings here? This is just 

```
filter (\(C c' t' :<= _) -> c == c' && matches t' pat) xs
```

Do you mean for the predicates? Like you want 

``` haskell
let sameClass = c == c' 
    matchesPat = matches t' pat
in (...)
```